Manage mirakl refunds

ablanchard opened the pull request:
Hi
I developped a new cron inside the connector to manage mirakl refund.
## Why
For our project stopcovid19.fr we needed to handle refunds created by Mirakl. Refunds can be created in Mirakl without stripe knowing and without the e-commerce platform knowing. So we needed "something" to do it. We choose to add a cron inside this connector because it was our easiest component to change. One more detail, mirakl refunds can be partial.
I'm proposing this cron to you to offer this refund management.
## How
With Mirakl, managing the refunds can be done using few APIs : 
 - Polling PA12 for pending refunds
 - Process those refunds in Stripe (refund the charge and reverse the transfer)
 - Confirm that the refund has been processed by calling PA02 (after this the refund will not be returned by PA12)
## Implementation
I followed the general architecture of the app: 
Mirakl refunds have their own table where we can track all the details of it. They can be access through a repository `MiraklRefundRepository` 
The entry point `ProcessRefundCommand` creates refunds in db and send a message on the bus. A new process can be trigger with `bin/console connector:dispatch:process-refund -e prod
`
The `ProcessRefundHandler` is in charge of the business logic I described above. 
The problem here is that we don't want to refund twice a charge. Even if our connector fails after the stripe refund creation and before the persist.
To protect against that, we check via stripe api the presence of a refund or of a reversal before calling the creation apis.
Before confirming the refund to mirakl, we want our connector to be in possession of both the stripe refund id and the stripe reversal id.
I implemented alos the `RefundFailedMessage` to have it in the failed queue later. I also added the failed refunds in the notification email destinated to the operator.
I added some test following what was already present.

Hope you guys, will like it. And don't hesitate to reach out on any problem of this PR.

kraynel commented on src/Handler/ProcessRefundHandler.php:
```diff
@@ -0,0 +1,196 @@
+            $this->processRefundInStripe($refund, $transfer);
+            $this->processReversalInStripe($refund, $transfer);
+            // Valid refund on mirakl PA02
+            $this->validateRefundInMirakl($refund);
```
If I understand correctly, you validate the refund on mirakl only if both refund + reverse transfer worked on Stripe. 
 If the seller already payed out, the reversal could fail, but the refund may be ok.

If the refund worked, the client will see the money, but the operator will not see the refund as successful in the interface.
Could it be a problem? The client could argue that they never received the refund.

Maybe we should validate the refund as soon as the refund is ok on Stripe, and warn the operator in case the reversal fails? @youssef-stripe what do you think?

youssef-stripe replied:
I agree with you Kevin. The refund should be marked as valid in Mirakl if `processRefundInStripe` succeeds. Then if processReversalInStripe fails, it should be handled the same as a Transfer failure: reverse transfer is later retried and eventually an alert is sent to the administrator.

The reason is that you wouldn't reverse the transfer if the refund fails (which it can) while the opposite is not true as you'd still want to refund the customer even if the transfer cannot be reversed immediately.

What do you think Alexandra?

ablanchard replied:
Hi, 
Yes, it will make more sense to confirm the refund in Mirakl once the refund has been done to the customer. 
In this case, I will change the way I handle transfer reversal and do it in the `ProcessTransferHandler`.

kraynel commented:
Thank you very much for your strong contribution Alexandre, and thank you for adding the tests!

ablanchard commented on tests/StripeMockedHttpClient.php:
```diff
@@ -79,6 +96,83 @@ private function getJsonStripePayout($payoutId)
+              "object"=> "list",
+              "data"=> [
+                [
+                    "id" => "refund_x",
```
I think `php-cs-fixer` doesn't look in test directory

ablanchard commented:
Hi,
I updated the `ProcessRefundHandler` to process the refund and the reversal separately. 
Once the refund is created in Stripe, the handler validate it in Mirakl. The reversal can occur later. 
To track if the refund has been validated in Mirakl, I added a column `mirakl_validation_time`. This way a `MiraklRefund` in the connector will be in state `CREATED` only if stripe refund, stripe reversal and Mirakl validation had been done.
